Add progress indicator feature

yash-chowdhary opened the pull request:
**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [x] New feature

<!--
    If this pull request is addressing an issue, link to the issue: "Fixes #xxx" or "Resolves #xxx"
-->
<!--
    Please ensure your pull request is ready:
    - Bug fix PR that is non-trivial **should** add a page or unit test for regression testing.
    - Feature PR **must** add a page to the user guide for demo.
    - Enhancement PR **should** update the user guide.

    Otherwise, prefix your PR title with "[WIP]".
-->

**What is the rationale for this request?**
This PR adds the progress indicator as outlined in #983. This gives authors the choice to add a reading progress indicator to pages.

**What changes did you make? (Give an overview)**
Added `.progress` CSS class in `markbind.css` that styles the progress bar. 
Added logic in `setup.js` to get the scroll percentage in a webpage.
In order to enable the progress bar in the pages, the following markup can be added to the header file:
```html
<div class="progress"></div>
```

**Provide some example code that this change will affect:**

<!-- Paste the example code below: -->
```js
function setupReadingProgress() {
  const h = document.documentElement;
  const b = document.body;
  const st = 'scrollTop';
  const sh = 'scrollHeight';
  const progress = document.querySelector('.progress');
  let scroll;

  document.addEventListener('scroll', () => {
    scroll = Number(h[st] || b[st]) / Number((h[sh] || b[sh]) - h.clientHeight);
    scrollPercentage = scroll * 100;
    progress.style.setProperty('--scroll', `${scrollPercentage}%`);
  });
}
```


**Testing instructions:**
* Run `markbind init` 
* Add the markup to the header file (`_markbind/headers/header.md`)
* Run `markbind serve` to generate the site
* Scroll to see the progress indicator

**Proposed commit message: (wrap lines at 72 characters)**

<!--
    See this link for more info on how to write a good commit message:
    https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message
-->
Add reading progress indicator
<!--
    Some of the responses that you gave to the previous questions might
    provide you with the information needed to craft your commit message.
-->

marvinchin commented on asset/js/setup.js:
```diff
@@ -95,6 +95,21 @@ function setupSiteNav() {
 }
 
+function setupReadingProgress() {
+  const h = document.documentElement;
```
Could we use more descriptive variable names here?

yash-chowdhary replied:
I've removed them altogether. They're not really required.

marvinchin commented on asset/js/setup.js:
```diff
@@ -95,6 +95,21 @@ function setupSiteNav() {
+  let scroll;
+
+  document.addEventListener('scroll', () => {
+    scroll = Number(h[st] || b[st]) / Number((h[sh] || b[sh]) - h.clientHeight);
```
Is there a reason why we need to use `Number` here?

marvinchin replied:
I think it might also be worth splitting this assignment up into several steps. There is a lot of logic going on here that is not immediately obvious.

yash-chowdhary replied:
In the previous iteration, I had set `scroll` to the expression which is now separated into 2 lines - defining `scroll` and `scrollPercentage`. This caused some lint issues with the operators. That's why I used `Number`. 
I have now removed the `Number` typecasting.
I have also switched to the `.` notation for accessing the properties.

marvinchin commented on asset/js/setup.js:
```diff
@@ -95,6 +95,21 @@ function setupSiteNav() {
+  const st = 'scrollTop';
+  const sh = 'scrollHeight';
+  const progress = document.querySelector('.progress');
+  let scroll;
```
Please correct me if I'm wrong, but could we make this a `const` within the event listener instead? We don't seem to need the previous scroll value to compute the new one.

yash-chowdhary replied:
Yes, you're right. :) Resolved.

marvinchin commented on asset/js/setup.js:
```diff
@@ -95,6 +95,21 @@ function setupSiteNav() {
+function setupReadingProgress() {
+  const h = document.documentElement;
+  const b = document.body;
+  const st = 'scrollTop';
```
Is there a reason we are assigning these constant field names to a variable? Why not just do `document.body.scrollTop` etc. instead? It might be more verbose, but I think that is more readable compared to using ultra-short variable names.

yash-chowdhary replied:
Resolved. :)

yash-chowdhary commented:
Thanks for your review, @marvinchin :) 
Refactored the code and added the progress bar to `test_site`.

marvinchin commented on asset/js/setup.js:
```diff
@@ -95,6 +95,18 @@ function setupSiteNav() {
+
+  document.addEventListener('scroll', () => {
+    const amountScrolled = (document.documentElement.scrollTop || document.body.scrollTop);
+    const pageHeight = ((document.documentElement.scrollHeight || document.body.scrollHeight)
```
Just a nit - I don't think you need the outermost brackets.

yamgent commented on docs/userGuide/syntax/headers.mbdf:
````diff
@@ -23,6 +23,18 @@ In the page that you want to include the header:
 ```
 </div>
 
+You can also choose to add a <trigger for="pop:progress-indicator">reading progress indicator</trigger> to let the reader know how far along they are in the page.
````
Due to how our current search works, your feature won't be discoverable if the authors attempt a search on our documentation website. I suggest giving a heading to your section:

`### Adding progress indicators`

yash-chowdhary replied:
Sure, will do 👍

marvinchin commented on asset/js/setup.js:
```diff
@@ -95,6 +95,18 @@ function setupSiteNav() {
+function setupReadingProgress() {
+  const progress = document.querySelector('.progress');
+
+  document.addEventListener('scroll', () => {
```
Oops, one more thing! We should only add the event listener when the progress indicator is on the page. Otherwise, `progress` might not be defined.
<img width="837" alt="Screenshot 2020-01-20 at 5 04 47 PM" src="https://user-images.githubusercontent.com/17447681/72712908-fccb1900-3ba6-11ea-932e-c8c89fe36634.png">

yash-chowdhary replied:
Fixed

openorclose commented:
This causes quite a big jump when the content height is changed (such as with panels or on resize).

Is it possible to respond to these events immediately rather than waiting for a scroll event to occur?
![prog](https://user-images.githubusercontent.com/3646725/72718413-ddd18480-3bb0-11ea-93af-9f6960a51cb0.gif)

yash-chowdhary commented:
Thanks for the observation @openorclose. :) 
I've added a hacky feature that will periodically check for document changes (as outlined [here](https://stackoverflow.com/questions/14866775/detect-document-height-change)), as there is no straightforward solution to checking this.
I've also added a listener for window resizing. 
Ready for review.

damithc commented:
@yash-chowdhary this is a very low-value feature IMO (note most other sites don't have such a feature). Therefore, make sure there is no performance cost on those who don't use the feature and the future code maintenance cost is VERY low. If that's not possible, the right thing to do is to shelve this feature.

yash-chowdhary commented:
> make sure there is no performance cost on those who don't use the feature

@damithc There are no performance issues for websites/pages that don’t use the progress indicator as the listeners are only added if and when the indicator is enabled.

damithc commented:
> @damithc There are no performance issues for websites/pages that don’t use the progress indicator as the listeners are only added if and when the indicator is enabled.

That's good. The amount of code changes in this PR is still a bit worrying though. Feels like the cost is not worth the benefit. All this code needs to be maintained in future. To clarify, I feel the value of the feature is low because the vertical scrollbar of the browser also serves as a similar indicator of progress through the page. I didn't object to this feature at first because I thought it is going to be simple to implement. What do you guys think?

ang-zeyu commented:
> What do you guys think?

I think we should keep it, in the future we could even extend this with something similar to https://bootstrap-vue.js.org/docs/directives/scrollspy/ to provide more value over the browser scrollbar.
( e.g. a popover that appears on scrolling indicating the current 'topic' being viewed, then fades away after a while so it dosen't affect readability )

> I've added a hacky feature that will periodically check for document height changes (as outlined [here](https://stackoverflow.com/questions/14866775/detect-document-height-change)), as I couldn't find a straightforward solution to this.
 
How about setting a simple transition duration for the width? Then instead of modifying `--scroll` you would simply modify the width. https://www.w3schools.com/howto/howto_js_scroll_indicator.asp

It's a compromise for performance though, since it would only smooth out the sharp transition once the user scrolls after opening a panel. But personally I would prefer this over the performance cost of `setTimeout`

One of my mice is rather hard to scroll, and I use a larger lines-per-scroll setting to compensate for that. It may help to even out the transition for such cases as well.

yash-chowdhary commented:
> How about setting a simple transition duration for the width? Then instead of modifying `--scroll` you would simply modify the width. https://www.w3schools.com/howto/howto_js_scroll_indicator.asp
 
I've tested it out with different values. This would help minimize the jump.

yash-chowdhary commented:
@marvinchin @yamgent any thoughts on how to proceed?

yamgent commented:
> @marvinchin @yamgent any thoughts on how to proceed?

Besides the vertical scrollbar cons that Prof has mentioned, the bar might also be confused with the content loading bar (e.g. the loading bar on the top that YouTube uses when navigating between pages). The content loading bar is also a common visual design on mobile phones. [This article](https://uxdesign.cc/pros-and-cons-of-progress-indicator-as-a-scroll-bar-345f19967cb6) provides some arguments about the pros and cons of a progress indicator.

I am inclined to agree with Prof that this feature might not see widespread use, so there's very little value in having this in.

yash-chowdhary commented:
Closing as a result of the above discussion.
